File: packages/protocol/test/sensitive.test.ts
/**
 * `docs/SECURITY.md` section 18 forbids raw credentials in logs. The enrolment
 * token and the session capability are the two credentials this protocol
 * carries, so every representation other than the deliberate wire encoding must
 * redact them.
 */

import assert from "node:assert/strict";
import test from "node:test";
import { inspect } from "node:util";

import { readFixture } from "../src/fixtures.ts";
import {
  decodeControlFrame,
  decodeDataStreamHeaderFrame,
  encodeControlFrame,
  encodeDataStreamHeaderFrame,
} from "../src/frame.ts";
import { REDACTED, SensitiveString } from "../src/sensitive.ts";

const SECRET = "MU5vbmNlLTAxSkFCQ0RFRkdISktMTU5PUFFSU1RVVg";

function registrationRequest() {
  const decoded = decodeControlFrame(readFixture({ file: "valid/registration-request.json" }));
  assert.ok(decoded.ok);
  assert.equal(decoded.value.type, "connector.registration.request");
  return decoded.value;
}

test("the enrolment token is redacted in every default representation", () => {
  const frame = registrationRequest();
  const token = frame.payload.enrolment_token;
  assert.equal(token.reveal(), SECRET, "the fixture no longer carries the expected token");

  const representations: Record<string, string> = {
    toString: token.toString(),
    "String()": String(token),
    template: `${token}`,
    concatenation: `` + token,
    "JSON.stringify(token)": JSON.stringify(token),
    "JSON.stringify(payload)": JSON.stringify(frame.payload),
    "JSON.stringify(frame)": JSON.stringify(frame),
    "util.inspect(token)": inspect(token),
    "util.inspect(payload)": inspect(frame.payload, { depth: null }),
    "util.inspect(frame)": inspect(frame, { depth: null }),
    "console.log format": inspect({ event: "registration", frame }, { depth: null }),
  };

  for (const [name, representation] of Object.entries(representations)) {
    assert.ok(!representation.includes(SECRET), `${name} leaked the enrolment token`);
    assert.ok(representation.includes(REDACTED), `${name} carries no redaction marker`);
  }
});

test("the canonical encoding carries the real enrolment token", () => {
  const frame = registrationRequest();
  const encoded = encodeControlFrame(frame);
  assert.ok(encoded.includes(SECRET), "the wire frame must carry the real enrolment token");
});

test("the session capability is redacted", () => {
  const decoded = decodeDataStreamHeaderFrame(readFixture({ file: "valid/data-stream-header.json" }));
  assert.ok(decoded.ok);
  const capability = decoded.value.session_capability.reveal();
  assert.ok(capability.length > 0);

  assert.ok(!JSON.stringify(decoded.value).includes(capability));
  assert.ok(!inspect(decoded.value, { depth: null }).includes(capability));
  assert.ok(encodeDataStreamHeaderFrame(decoded.value).includes(capability));
});

test("a sensitive string reports its length without revealing its value", () => {
  const secret = new SensitiveString("abcdef");
  assert.equal(secret.length, 6);
  assert.equal(secret.toString(), REDACTED);
  assert.equal(secret.reveal(), "abcdef");
});

File: packages/protocol/connectorv1/sensitive_test.go
package connectorv1

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log/slog"
	"strings"
	"testing"
)

// docs/SECURITY.md section 18 forbids raw credentials in logs. The enrolment
// token and the session capability are the two credentials this protocol
// carries, so every representation other than the deliberate wire encoding must
// redact them.

const secret = "MU5vbmNlLTAxSkFCQ0RFRkdISktMTU5PUFFSU1RVVg"

func decodedRegistrationRequest(t *testing.T) (Frame, RegistrationRequest) {
	t.Helper()
	frame, failure := DecodeControlFrame(readFixture(t, "valid/registration-request.json"))
	if failure != nil {
		t.Fatalf("fixture refused: %v", failure)
	}
	request, ok := frame.Payload.(RegistrationRequest)
	if !ok {
		t.Fatalf("unexpected payload %T", frame.Payload)
	}
	return frame, request
}

func TestEnrolmentTokenIsRedactedInEveryDefaultRepresentation(t *testing.T) {
	_, request := decodedRegistrationRequest(t)
	if request.EnrolmentToken.Reveal() != secret {
		t.Fatalf("fixture no longer carries the expected token")
	}

	representations := map[string]string{
		"String":            request.EnrolmentToken.String(),
		"GoString":          request.EnrolmentToken.GoString(),
		"fmt %v":            fmt.Sprintf("%v", request.EnrolmentToken),
		"fmt %s":            fmt.Sprintf("%s", request.EnrolmentToken),
		"fmt %q":            fmt.Sprintf("%q", request.EnrolmentToken),
		"fmt %x":            fmt.Sprintf("%x", request.EnrolmentToken),
		"fmt %#v":           fmt.Sprintf("%#v", request.EnrolmentToken),
		"struct %v":         fmt.Sprintf("%v", request),
		"struct %+v":        fmt.Sprintf("%+v", request),
		"struct %#v":        fmt.Sprintf("%#v", request),
		"frame %+v":         fmt.Sprintf("%+v", mustDecodedFrame(t)),
		"error interpolate": fmt.Errorf("registration failed for %v", request).Error(),
	}
	for name, representation := range representations {
		if strings.Contains(representation, secret) {
			t.Errorf("%s leaked the enrolment token: %s", name, representation)
		}
		if !strings.Contains(representation, Redacted) {
			t.Errorf("%s did not carry the redaction marker: %s", name, representation)
		}
	}
}

func mustDecodedFrame(t *testing.T) Frame {
	t.Helper()
	frame, _ := decodedRegistrationRequest(t)
	return frame
}

func TestEnrolmentTokenIsRedactedByEncodingJSON(t *testing.T) {
	_, request := decodedRegistrationRequest(t)

	marshalled, err := json.Marshal(request)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	if bytes.Contains(marshalled, []byte(secret)) {
		t.Fatalf("encoding/json leaked the enrolment token: %s", marshalled)
	}
	if !bytes.Contains(marshalled, []byte(Redacted)) {
		t.Fatalf("encoding/json did not redact the enrolment token: %s", marshalled)
	}
}

func TestEnrolmentTokenIsRedactedBySlog(t *testing.T) {
	_, request := decodedRegistrationRequest(t)

	var sink bytes.Buffer
	logger := slog.New(slog.NewJSONHandler(&sink, nil))
	logger.Info("connector registration", "token", request.EnrolmentToken, "request", request)
	if strings.Contains(sink.String(), secret) {
		t.Fatalf("slog leaked the enrolment token: %s", sink.String())
	}
	if !strings.Contains(sink.String(), Redacted) {
		t.Fatalf("slog did not redact the enrolment token: %s", sink.String())
	}
}

// The wire encoding is the one place the secret is deliberately revealed.
func TestCanonicalEncodingCarriesTheRealToken(t *testing.T) {
	frame, _ := decodedRegistrationRequest(t)
	encoded, err := EncodeControlFrame(frame)
	if err != nil {
		t.Fatalf("encode: %v", err)
	}
	if !bytes.Contains(encoded, []byte(secret)) {
		t.Fatal("the wire frame must carry the real enrolment token")
	}
}

func TestSessionCapabilityIsRedacted(t *testing.T) {
	header, failure := DecodeDataStreamHeaderFrame(readFixture(t, "valid/data-stream-header.json"))
	if failure != nil {
		t.Fatalf("fixture refused: %v", failure)
	}
	capability := header.SessionCapability.Reveal()
	if capability == "" {
		t.Fatal("fixture no longer carries a capability")
	}

	rendered := fmt.Sprintf("%+v", header)
	if strings.Contains(rendered, capability) {
		t.Fatalf("the data-stream header leaked the session capability: %s", rendered)
	}

	marshalled, err := json.Marshal(header)
	if err != nil {
		t.Fatalf("marshal: %v", err)
	}
	if bytes.Contains(marshalled, []byte(capability)) {
		t.Fatalf("encoding/json leaked the session capability: %s", marshalled)
	}

	encoded, err := EncodeDataStreamHeaderFrame(header)
	if err != nil {
		t.Fatalf("encode: %v", err)
	}
	if !bytes.Contains(encoded, []byte(capability)) {
		t.Fatal("the wire header must carry the real capability")
	}
}
